lib/Sema/TypeChecker.cpp

```diff
@@ -388,6 +388,10 @@ Type swift::performTypeResolution(TypeRepr *TyR, ASTContext &Ctx,
         // FIXME: Don't let unbound generic types escape type resolution.
         // For now, just return the unbound generic type.
         return unboundTy;
+      }, /*placeholderHandler*/ [&]() {
+        // FIXME: Don't let placeholder types escape type resolution.
+        // For now, just return the placeholder type.
+        return Ctx.ThePlaceholderType;
```

Comment:
Maybe not something to do here specifically, but I was really hoping we could use placeholder types to make UnboundGenericType obsolete. Wherever type resolution forms one, it could instead build a bound generic type full of placeholders. With that in mind, having them escape unopened is not so bad.

Reply:
Yeah this was on my mind as well, especially since one of the goals of the proposal is to have `UnboundGeneric` be semantically equivalent to `UnboundGeneric<_, ..., _>`. Obsoleting `UnboundGenericType` would really help validate that approach.

It's been a bit since I worked on this but IIRC I started down that path, and then realized there was enough complexity (and bits that I didn't quite understand) in `ConstraintSystem::openUnboundGenericType` that it was going to turn into a bit of a rabbit hole, so I punted on that in favor of getting the basic functionality working.

Thank you for taking a look at this!

Reply:
> Maybe not something to do here specifically, but I was really hoping we could use placeholder types to make UnboundGenericType obsolete. Wherever type resolution forms one, it could instead build a bound generic type full of placeholders. With that in mind, having them escape unopened is not so bad.